### PR TITLE
Feature/221 serialize enums as flattened structs

### PR DESCRIPTION
**Pull request by raj-nimble**
This MR isn't complete, still need to add the correct functionality to array builder / sequence builder.
This is just as an example of the direction I was going, was hoping for your thoughts?

With this change, `from_samples` outputs Fields with the desired flattened structure.

### CODE REVIEW

File: serde_arrow/src/internal/schema/from_samples/mod.rs [hunk 1]
```diff
@@ -694,4 +702,131 @@ mod test {
...
+                Number::Real { value: 1.0 },
+                Number::Complex { i: 0.5, j: 0.5 },
+            ],
+            TracingOptions::default(),
+            expected,
+        );
+    }
+
+    #[test]
+    fn example_enum_as_struct() {
```
Comment (raj-nimble) [inline serde_arrow/src/internal/schema/from_samples/mod.rs:750]: should probably combine this test and the following one to check for equality between the generated fields using both the enum and the struct

**Comment by chmp** [comment]
Hi @raj-nimble 

Thanks. I like the idea a lot!

ATM, I don't really have the time for serde arrow. I will have a proper look next week! 

First feedback: it's great, that you target the 0.12 release. I would like to limit the current main to bug fixes for the 0.11 release. 

Re builders we can also figure it out together, once I'm back at my laptop. I imagine the deserialization logic will be a bit tricky. Maybe the  impl of untagged enums in serde could be an inspiration (I like use the rust playground to expand the generated deserialize impl).

**Comment by raj-nimble** [comment]
Hi @chmp that's great, I'm happy you like it. I will continue to iterate slowly on my own this week and I look forward to discussing it more with you next week.

**Review by chmp — COMMENTED** [review]
I left some comments and thoughts. Thinking about next steps: 

For serialization you should be able to follow the `StructArrayBuilder`, maybe even implement the `ArrayBuilder` as a collection of `StructBuilders` and only merge the results in `to_array`. Most likely having some annotation with a strategy is required to allow selecting a specialized implementation.

For deserialization I would implement the strategy to detect any non-null field, select the relevant variant and then follow the impl of the standard `EnumDeserializer`. However, I think getting this logic right, will be really tricky. Probably, I would implement this as serialization only for now.

**Comment by raj-nimble** [comment]
Hi @chmp . Thanks so much for the comments! I haven't made any progress since last week as I was working on something else, but with your comments now I can hopefully circle back to this starting next week and I'll update you as soon as I have made some progress. Thanks!

File: serde_arrow/src/internal/schema/tracer.rs [hunk 2]
```diff
@@ -965,18 +965,46 @@ impl UnionTracer {
...
+            // For each variant with name variant_name
+            // For each variant_field with field_name
+            // Add field {variant_name}_{field_name} -> variant_field.to_field() that is nullable
+
+            for variant in &self.variants {
+                if let Some(variant) = variant {
+                    // TODO: does this break if there are no child fields?
+                    let schema = variant.tracer.to_schema()?;
+                    for mut field in schema.fields {
+                        field.name = format!("{}_{}", variant.name.to_lowercase(), field.name);
```
[the remark this answers is not in the document]
Comment (raj-nimble) [reply serde_arrow/src/internal/schema/tracer.rs:983]: I like your suggestion about using a separator that is not valid Rust code to guarantee certain cases cannot arise.

[the remark this answers is not in the document]
Comment (raj-nimble) [reply serde_arrow/src/internal/schema/tracer.rs:983]: > What happens with only nullable fields in variant? (I would throw an error)
> * Most likely your deserialization logic will look for any-null field and select the matching variant. However, if there is a variant for which all fields are `None` this logic will fail.

For this case, I believe it's best to continue to serialize it. I agree at first we simply don't add support for deserialization and it seems perfectly fine for the user to serialize out an enum where the variant has all nullable fields. If we want to indicate an error, we can check all fields null when deserializing (whenever we get there) and raise the error at that point.

**Comment by raj-nimble** [comment]
> Probably, I would implement this as serialization only for now.
Agreed, it's best to implement the feature as serialization only for now. Can add a separate issue to track adding deserialization.

File: serde_arrow/src/internal/schema/tracer.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (raj-nimble) [reply serde_arrow/src/internal/schema/tracer.rs:983]: Hi @chmp , just to clarify on the points you made and make sure nothing is missed

> * What happens if there is a conflict between unions (see below)?

I have changed the implementation to use the `::` separator which I believe addresses the chance of a inadvertent naming collision.

> * What happens with unions without fields (I would throw an error, as it will be hard to detect the correct variant on deserialization)?

This currently fails in the `to_field()` method of the UnionTracer when it detects enums without data.
 
> * What happens with non-struct variants (unit variants, tuple variants, new-type variants?) (I would probably forbid unit variants, as they fall under the no-fields rule)

- unit variants: if `enums_without_data_as_strings()` is enabled, fields are serialized using that approach. if not enabled, it throws an error, as previously done.
- tuple variants: This currently re-raises the error that occurs when `to_schema` is called on the variant tracer in the `UnionTracer::to_field()` method. I believe this maintains the existing behavior?
- new-type variants: this should be exactly the same as tuple structs, right? I believe it's handled in the same way?

> * What happens with only nullable fields in variant? (I would throw an error)

I commented on this separately above

> It would be great if you could add tests for these corner cases.

I added some tests confirming what I hope you agree is the expected behavior but if I misunderstood anything, please feel free to point it out. Thanks!

File: serde_arrow/src/internal/schema/tracer.rs [hunk 3]
```diff
@@ -965,18 +965,46 @@ impl UnionTracer {
...
+                if let Some(variant) = variant {
+                    fields.push((i8::try_from(idx)?, variant.tracer.to_field()?));
+                } else {
+                    fields.push((i8::try_from(idx)?, unknown_variant_field()));
+                };
+            }
+            data_type = DataType::Union(fields, UnionMode::Dense);
         }
 
         Ok(Field {
```
[the remark this answers is not in the document]
Comment (raj-nimble) [reply serde_arrow/src/internal/schema/tracer.rs:1005]: Indeed it does! Added.

File: serde_arrow/src/internal/schema/from_samples/mod.rs [hunk 4]
```diff
@@ -694,4 +695,47 @@ mod test {
...
+                complex_i: Some(0.5),
+                complex_j: Some(0.5),
+                ..Default::default()
+            },
+        ];
+
+        let opts = TracingOptions::default().enums_with_data_as_structs(true);
+
+        let enum_tracer = Tracer::from_samples(&enum_items, opts).unwrap();
+        let struct_tracer = Tracer::from_samples(&struct_items, TracingOptions::default()).unwrap();
```
[the remark this answers is not in the document]
Comment (raj-nimble) [reply serde_arrow/src/internal/schema/from_samples/mod.rs:734]: I added a basic UT just to see if it works but I want to move the tests together to share the utility code and test all edge cases with both converters.

**Comment by raj-nimble** [comment]
Hi @chmp ,

I'm struggling to find an elegant way to do the serialization.
I started out attempting to modify StructBuilder to handle it but the issue becomes that the value being serialized is the actual Enum and it gets quite messy to attempt to detect that in ArrayBuilder, and deconstruct it into it's fields to call serialize on that.

I considered redoing the schema creation to continue to use UnionBuilder and then add logic there to serialize it as a struct instead but I think that still ran into problems with arrow writer.

I then considered creating an entirely new builder, something like an `EnumStructBuilder` and I will probably spend a bit more time there. I think that would move the complexity into OuterSequenceBuilder::new, into the `build_builder` and `build_struct` inner methods but I'm not quite sure yet.

Do you have any time to take a closer look and give some guidance on exactly how this might best be done?

Thanks,
Raj

**Comment by chmp** [comment]
@raj-nimble  good questions, what's the best option. 


One option would probably be to copy the code of `UnionBuilder` into a separate file and adapt it:

- build a union builder first, then modify it. This should fail if any variant is not serialized as a struct (as otherwise it does not make sense to merge it). I would include the logic merging the field names here.
- merge the different struct builders in `into_array()`. Note that this would require to call `push_default` or `push_none` on the unused builders

In pseudo code:

```rust
    // probably you need to do something different here, as you plan to use `Struct` as the type with a custom strategy
    pub fn from_union_builder(builder: UnionBuilder) -> Result<FlattenedUnionBuilder> {
        let builders = Vec::new();
        for (child_builder, child_meta) in builder.fields {
           let ArrayBuilder::Struct(child_builder) = child_builder else { fail!("Children must be structs") };
           // modify the fields to include the variant prefix, the variant name will be in child_meta.name
        }

        // construct the builder
       Ok(FlattenedUnionBuilder { ... })
    }

    pub fn into_array(self) -> Result<Array> {
        let mut fields = Vec::new();
        
        // note: the meta data for the variant is most likely not used, I would simply drop it
        for (_, builder) in self.fields.into_iter().enumerate() {
            let ArrayBuilder::Struct(builder) = builder else { fail!("..."); };

            // concatenate the fields of the different variants
            for (field, meta) in builder.fields {
                fields.push((idx.try_into()?, builder.into_array()?, meta));
             }
        }

        Ok(Array::Struct(StructArray {
            // note: you will most likely need to keep track of the number of elements being added, simply add self.len += 1 or similar in the different `serialize_*` methods
            len: self.len,
            validity: self.seq.validity,
            fields,
        }))
    }

    pub fn serialize_variant(&mut self, variant_index: u32) -> Result<&mut ArrayBuilder> {
        self.len += 1;

        let variant_index = variant_index as usize;
        
        // call push_none for any variant that was not selected
        for (idx, builder) in self.fields.iter_mut().enumerate() {
           if idx != variant_idx {
               builder.serialize_none()?;
            }
         }
        
        let Some(variant_builder) = self.fields.get_mut(variant_index) else {
            fail!("Could not find variant {variant_index} in Union");
        };

        Ok(variant_builder)
    }
```

**Comment by raj-nimble** [comment]
Hi @chmp , thanks so much for your suggestions / guidance, it was immensely helpful.
I have pushed up the current draft of work. I haven't written any unit tests or done any cleanup up yet but this version is working with the new builder and moreover, I tested it with our application and happy path works which is a good sign. I plan to spend the next day or two cleaning it up and writing some tests.

**Comment by raj-nimble** [comment]
I commented here on my current workaround for UUIDs when using `from_type`
https://github.com/chmp/serde_arrow/issues/203#issuecomment-2362294959

**Comment by chmp** [comment]
@raj-nimble regarding tests: I will add a method to easily construct the new internal array objects. this way you should be able to easily get the underlying data out and compare whether everything worked.

File: serde_arrow/src/internal/serialization/flattened_union_builder.rs [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        try_(|| {
+            let variant_builder = self.serialize_variant(variant_index)?;
+            variant_builder.serialize_struct_start(variant, len)?;
+            Ok(variant_builder)
+        })
+        .ctx(&ctx)
+    }
+}
+
+#[cfg(test)]
```
Comment (chmp) [inline serde_arrow/src/internal/serialization/flattened_union_builder.rs:122]: So far my convention has been to move these test into the `serde_arrow/src/test_with_arrow` tree.  As written in the top-level comment, I will add support to simplify getting the data of the arrays so it will be easier to check the contents.

File: serde_arrow/src/internal/serialization/outer_sequence_builder.rs [hunk 6]
```diff
@@ -226,7 +227,52 @@ fn build_builder(path: String, field: &Field) -> Result<ArrayBuilder> {
...
+        T::Struct(children) => {
+            if let Some(Strategy::EnumsWithNamedFieldsAsStructs) =
+                get_strategy_from_metadata(&field.metadata)?
+            {
+                let mut related_fields: BTreeMap<&str, Vec<Field>> = BTreeMap::new();
+                let mut builders: Vec<(ArrayBuilder, FieldMeta)> = Vec::new();
+
+                for field in children {
+                    let Some(variant_name) = field.union_variant_name() else {
+                        // TODO: failure message
```
Comment (chmp) [inline serde_arrow/src/internal/serialization/outer_sequence_builder.rs:239]: probably better to use [`todo!()`](https://doc.rust-lang.org/stable/std/macro.todo.html) in these cases. This way you can easily search for it and still have the correct runtime behavior.

File: serde_arrow/src/internal/serialization/outer_sequence_builder.rs [hunk 7]
```diff
@@ -226,7 +227,52 @@ fn build_builder(path: String, field: &Field) -> Result<ArrayBuilder> {
...
+                    let mut new_field = field.clone();
+                    new_field.name = field_name;
+
+                    related_fields
+                        .entry(variant_name)
+                        .or_default()
+                        .push(new_field);
+                }
+
+                for (variant_name, fields) in related_fields {
```
Comment (chmp) [inline serde_arrow/src/internal/serialization/outer_sequence_builder.rs:257]: Why do you use two loops here? couldn't you simply build the builder where you push into `related_fields`? This way you also don't need to clone the fields.

File: serde_arrow/src/internal/serialization/simple_serializer.rs [hunk 8]
```diff
@@ -169,7 +169,7 @@ pub trait SimpleSerializer: Sized + Context {
     fn serialize_struct_start(&mut self, name: &'static str, len: usize) -> Result<()> {
         fail!(
             in self,
-            "serialize_start_start is not supported",
+            "serialize_struct_start is not supported",
```
Comment (chmp) [inline serde_arrow/src/internal/serialization/simple_serializer.rs:172]: Good catch. Thanks!

**Review by chmp — COMMENTED** [review]
Looks great so far!

**Comment by chmp** [comment]
@raj-nimble I started to work o a proper release for the 0.12 series. Hence the change in base branch.

File: serde_arrow/src/internal/serialization/outer_sequence_builder.rs (hunk 7 above)
Comment (raj-nimble) [reply serde_arrow/src/internal/serialization/outer_sequence_builder.rs:257]: I have at aggregate/collect all the fields for that variant first, before creating the builder.

File: serde_arrow/src/internal/serialization/outer_sequence_builder.rs (hunk 6 above)
Comment (raj-nimble) [reply serde_arrow/src/internal/serialization/outer_sequence_builder.rs:239]: fixed

File: serde_arrow/src/internal/serialization/flattened_union_builder.rs (hunk 5 above)
Comment (raj-nimble) [reply serde_arrow/src/internal/serialization/flattened_union_builder.rs:122]: sounds good, will move the tests once the helper methods are added

**Comment by raj-nimble** [comment]
Hi @chmp sorry for the delay here. I got pulled away from this for a bit but I'm really hoping to finish up here this week or early next week. Did you get a chance yet to add support to simplify getting the data of the arrays so I can move the tests?

**Comment by chmp** [comment]
Great to hear!

Re.: getting the data out of the arrays: you can use `build_arrays` of the `ArrayBuilder` and then match on the variants of the [`Array` enum](https://github.com/chmp/serde_arrow/blob/839d34b9a89bfe04f6c2f07996ff3234f26c734d/serde_arrow/src/internal/arrow/array.rs#L10). Something along the lines of

```rust
// see also: https://github.com/chmp/serde_arrow/blob/main/serde_arrow/src/test_with_arrow/impls/chrono.rs#L659
let mut builder = ArrayBuilder::new(schema).unwrap();
builder.extend(&items).unwrap();

let arrays = builder.build_arrays().unwrap();
let [array] = arrays.try_into().unwrap();

let Array::Struct(array) = array else { 
    panic!("Expected struct array, got: {array:?}"); 
};

let (first_field, meta) = &array.fields[0];
assert_eq!(meta.name, "...");
assert_eq!(first_field, Array::Int32(PrimtiveArray { validity: None, values: vec![1, 2, 3, 4] }));
```

**Comment by raj-nimble** [comment]
Hi @chmp I have an issue and I am not yet certain how to fix it. Maybe you can provide some insight?

The error I am getting is:
```
thread 'test_with_arrow::impls::flattened_union::test_flattened_union_with_nested_enum' panicked at serde_arrow/src/test_with_arrow/impls/flattened_union.rs:230:10:
failed to serialize: Error: Cannot push null for non-nullable array (data_type: "UInt32", field: "$.data.Two.opts.loc.key")
```
To repro/demonstrate, lets say we have this nested struct->enum->struct->enum structure
```
#[derive(Serialize, Deserialize)]
struct ComplexMessage {
    data: MsgData,
}

#[derive(Serialize, Deserialize)]
enum MsgData {
    One { data: usize },
    Two { opts: MsgOptions },
}

#[derive(Serialize, Deserialize)]
struct MsgOptions {
    loc: Location,
}

#[derive(Serialize, Deserialize)]
enum Location {
    Left,
    Right,
}
```
and we have a schema and data as follows
```
#[derive(Serialize, Deserialize)]
struct ComplexMessage {
    data: MsgData,
}

#[derive(Serialize, Deserialize)]
enum MsgData {
    One { data: usize },
    Two { opts: MsgOptions },
}

#[derive(Serialize, Deserialize)]
struct MsgOptions {
    loc: Location,
}

#[derive(Serialize, Deserialize)]
enum Location {
    Left,
    Right,
}
```
the error above shows up when we serialize
```
    let mut builder = ArrayBuilder::new(nested_enum_schema()).unwrap();
    let serializer = Serializer::new(builder);
    nested_enum_data()
        .serialize(serializer)
        .expect("failed to serialize")
        .into_inner()
        .to_arrow()
        .expect("failed to serialize to arrow");
```
The fields look like this
```
[
    Struct(
        StructArray {
            len: 0,
            validity: None,
            fields: [
                (
                    UInt64(
                        PrimitiveArray {
                            validity: Some(
                                [],
                            ),
                            values: [],
                        },
                    ),
                    FieldMeta {
                        name: "One::data",
                        nullable: true,
                        metadata: {},
                    },
                ),
                (
                    Struct(
                        StructArray {
                            len: 0,
                            validity: Some(
                                [],
                            ),
                            fields: [
                                (
                                    Dictionary(
                                        DictionaryArray {
                                            indices: UInt32(
                                                PrimitiveArray {
                                                    validity: None,
                                                    values: [],
                                                },
                                            ),
                                            values: LargeUtf8(
                                                BytesArray {
                                                    validity: None,
                                                    offsets: [
                                                        0,
                                                    ],
                                                    data: [],
                                                },
                                            ),
                                        },
                                    ),
                                    FieldMeta {
                                        name: "loc",
                                        nullable: false,
                                        metadata: {},
                                    },
                                ),
                            ],
                        },
                    ),
                    FieldMeta {
                        name: "Two::opts",
                        nullable: true,
                        metadata: {},
                    },
                ),
            ],
        },
    ),
]
```
As I understand it, what is happening is:
* The first variant in the first data structure is serialized, and 2nd variant builder is skipped, so we call `StructBuilder::serialize_none()`
* `StructBuilder::serialize_none()` calls `DictionaryUTF8Builder::serialize_default()`
* `DictionaryUTF8Builder::serialize_default()` calls `IntBuilder::serialize_none()`
* `IntBuilder::serialize_none()` calls `self.array.push_scalar_none()` which calls `set_validity()` with no buffer and value false which fails

So while I understand how we arrive at that section of the code, I don't really understand the underlying behavior of the Dict builder or Int Builder on why this is so. I believe the Dict Builder is a result of the `TracingOptions::enums_without_data_as_strings` field which we do want but I'm not sure how to approach allowing this variant to be none.

Do you have any thoughts on what I might need to do here?

**Comment by chmp** [comment]
@raj-nimble Oh that's a tricky. The reason is the following: 

- A null struct value needs to have value for its fields, that why `serialize_none` calls `serialize_default` for each of its fields
- The call to `serialize_default` in the dictionary array serializes a null value, as there is no reasonable default. The default for strings, an empty string, is not guaranteed to exist. The index `0` is not guaranteed to map to a valid string, if there are no values for this dictionary (probably a corner case, but nonethless)

One fix would be to force nested dictionary fields to be nullable with `enums_with_named_fields_as_structs`, regardless of the data.  Something like:

```rust
fn fix_dictionaries(field: &mut Field) {
  if matches!(field.data_type, DataType::Dictionary(_, _, _)) {
    field.nullable = true;
  } else if let DataType::Struct(children) = &mut field.data_type {
    for child in children {
       fix_dictionaries(child);
    }
  }
}
```

The recursion on struct is necessary, as nested struct that contain dictionaries will trigger the same error.

Just a heads up: unions in arrow are never nullable. For nested enums, that cannot be mapped to dictionary, you will run into a similar issue. Unfortunately, I have no idea what would be a solution here. I would suggest, to simply catchs this case and raise an error.

**Comment by raj-nimble** [comment]
Hi @chmp , sorry for the recent radio silence and lack of progress. I got pulled into more urgent tasks but this has definitely not fallen off my radar. I do plan to return to this hopefully in the next month.

For a summary of where I left off:
* I think most of it is good to go. The naive happy-path behavior is all working and some extended testing in our main application was all working as expected. There are some limitations but they were all expected and able to be handled.
* I think there is probably some gaps in unit test coverage still.
* I had found one specific bug that I was investigating at the time I dropped off. I have a unit test that reproduces the issue but I had not started investigating it yet.
* My plan when I can come back to this is to fix that bug, confirm with our application, add any other possibly missing unit tests, then open this up for official review.

Sorry for the delay with this.

**Comment by chmp** [comment]
@raj-nimble Thanks for the update. And no worries. I myself are quite busy with other things and can't really devote much time to serde_arrow. Just give me a ping when I should support :)
